src/fromager/metrics.py

```diff
@@ -0,0 +1,77 @@
+import functools
+import logging
+import time
+import typing
+from collections import defaultdict
+from datetime import timedelta
+
+from packaging.requirements import Requirement
+from packaging.version import Version
+
+_time_store: dict[str, dict[str, float]] = defaultdict(dict[str, float])
+_time_description_store: dict[str, str] = {}
+
+
+def timeit(description: str):
+    def timeit_decorator(func: typing.Callable) -> typing.Callable:
+        _time_description_store[func.__name__] = description
+
+        @functools.wraps(func)
+        def wrapper_timeit(
+            *, req: Requirement | None = None, **kwargs: typing.Any
+        ) -> typing.Any:
+            start = time.perf_counter()
+            ret = func(req=req, **kwargs)
+            end = time.perf_counter()
+            # get the logger for the module from which this function was called
+            logger = logging.getLogger(func.__module__)
+            version = (
+                kwargs.get("version")
+                or kwargs.get("dist_version")
+                or kwargs.get("resolved_version")
+            )
+            if not version:
+                version = _extract_version_from_return(ret)
```

Comment:
Which function needs this hack?

Reply:
any of the version resolvers like this one: https://github.com/python-wheel-build/fromager/blob/main/src/fromager/sources.py#L73